### PR TITLE
Add support for weeks in iso8601 duration parser (#350)

### PR DESCRIPTION
**Pull request by mlovic**
According to the ISO 8601, there are two formats for representing a duration:

> In both basic and extended format the complete representation of the expression for duration shall be [PnnW] or [PnnYnnMnnDTnnHnnMnnS].

This PR implements the first of those, as requested in #350.

    iex(1)> Timex.Duration.parse("P2W")
    {:ok, #<Duration(P14D)>}

However, I'm unsure about how to handle mixing the formats. While the ISO standard does not explicitly prohibit combining the two formats, it does clearly distinguish them, suggesting that they should be used separately. Because of this, I am inclined to treat any combination of the two formats as invalid input. However, both the Nodejs library linked to in the issue and Python's `isodate` (to give another example), _do_ support this, in different ways:

The node library will ignore the `W` designator, or everything except `W`, depending on which comes first.

The Python library will take into account both. From the library's test suite:

> ```
> (parse_duration("P1Y1M5W"),
>    [...should equal the difference between...]
>    date(2000, 1, 30),
>    date(2001, 4, 4)),
> ```

I also added some typespecs for the internal parsing functions in a different commit. It took me a while to understand those, so I thought the specs could be helpful. I'd be happy to remove them though.

### Checklist

- [x] New functions have typespecs, changed functions were updated
- [x] Same for documentation, including moduledocs
- [x] Tests were added or updated to cover changes
- [ ] Commits were squashed into a single coherent commit

### CODE REVIEW

**Comment by bitwalker** [comment]
This is great! Thanks for the PR!

The key is in the language used to describe the "basic" vs "extended" formats:

> In both basic and extended format the complete representation of the expression for duration shall be [PnnW] or [PnnYnnMnnDTnnHnnMnnS].

As somebody who is deep into parser theory and implementation, to me this reads that the "complete representation" of a duration in basic format is `PnnW`, and the "complete representation" of a duration in extended format is `PnnYnnMnnDTnnHnnMnnS` and that mixing the two should be considered invalid according to the spec (as implied by the "or"). I suspect Node's parser simply parses a "valid" format, either basic or extended, until it has fully parsed one of the two formats, ignoring what's left. Python's behaviour is interesting (even convenient!), but I believe it to be effectively an extension of the spec, and I would prefer to avoid going down that road in Timex.

File: lib/parse/duration/parsers/iso8601.ex [hunk 1]
```diff
@@ -54,6 +60,8 @@ defmodule Timex.Parse.Duration.Parsers.ISO8601Parser do
             {true, Duration.add(d, Duration.from_minutes(m))}
           {?S, s}, {true, d} ->
             {true, Duration.add(d, Duration.from_seconds(s))}
+          {?W, _w}, {_, _} ->
+            {:error, "W cannot be used with other time period designators"}
```
Comment (bitwalker) [inline lib/parse/duration/parsers/iso8601.ex]: Perhaps the error will be more helpful if we phrase it like:

```elixir
{?W, _w}, {_, _} ->
  {:error, "Found 'W', a basic format designator, but the parse indicates " <>
               "this is an extended format, mixing the two formats is disallowed"}
```

Comment (mlovic) [reply lib/parse/duration/parsers/iso8601.ex]: :+1: I've amended the commit

**Comment by bitwalker** [comment]
Thanks!

**Pull request merged**